R/filter-nm-data.R

```diff
@@ -0,0 +1,221 @@
+#' Helper for reading in and parsing the `$DATA` record.
+#'
+#' @note This cannot be used for _modifying_ a `$DATA` record.
+#' @noRd
+#' @keywords internal
+read_data_record <- function(.mod){
+  data_recs <- get_records(.mod, "data")
+  n_data <- length(data_recs)
+  if(n_data !=1){
+    recs_fmt <- purrr::map_chr(data_recs, function(rec) rec$format())
+    rlang::abort(
+      c(
+        glue::glue("Expected a single data record, but found {n_data}:\n\n"),
+        recs_fmt
+      )
+    )
+  }
+  data_recs[[1]]$parse()
+  return(data_recs[[1]])
+}
+
+
+#' Extract IGNORE or ACCEPT options from a data record
+#' @inheritParams filter_nm_data
+#' @noRd
+get_data_filter_exprs <- function(.mod){
+
+  data_rec <- read_data_record(.mod)
+
+  # Extract & format IGNORE/ACCEPT options
+  ignore_opts <- purrr::keep(data_rec$values, function(val){
+    inherits(val, "nmrec_option_value") && identical(val[["name"]], "ignore")
+  })
+
+  accept_opts <- purrr::keep(data_rec$values, function(val){
+    inherits(val, "nmrec_option_value") && identical(val[["name"]], "accept")
+  })
+
+  has_ignore <- !rlang::is_empty(ignore_opts)
+  has_accept <- !rlang::is_empty(accept_opts)
+
+  if(has_ignore && has_accept){
+    # Identical to NMTRAN error message if both are used
+    rlang::abort("ACCEPT list and IGNORE list may not both be used")
+  }else{
+    type <- dplyr::case_when(
+      has_ignore ~ "ignore",
+      has_accept ~ "accept",
+      # NA type will escape filtering
+      TRUE ~ NA_character_
+    )
+  }
+
+  # Pull out filters (remove parentheses and any quoting)
+  #  - can concatenate options since one or both are assumed to be empty
+  exprs <- purrr::map_chr(c(ignore_opts, accept_opts), function(val){
+    gsub("\\(|\\)", "", unquote_filename(val$value))
+  })
+
+  # Separate any (list) type expressions
+  exprs <- unlist(stringr::str_split(exprs, ",")) %>% stringr::str_trim()
+
+  return(
+    list(
+      type = type,
+      exprs = exprs
+    )
+  )
+}
+
+#' Function to translate NONMEM operators to R operators
+#' @noRd
+#' @keywords internal
+translate_nm_operator <- function(expr) {
+  # EQN and NEN are available after NONMEM 7.3
+  expr <- gsub("\\.EQ\\.|\\.EQN\\.", "==", expr)
+  expr <- gsub("\\.NE\\.|\\.NEN\\.|/=", "!=", expr)
+  expr <- gsub("\\.LT\\.", "<", expr)
```

Comment:
More literal strings; you can can leave them as is, but no need to do matching here. Just commenting for posterity. You can convert the first two (lines 76 and 77) to 5 literal `gsubs` and it will read clearer and run faster (just a touch faster, but it'll be faster).